app/services/course/assessment/answer/programming_auto_grading_service.rb

```diff
@@ -54,32 +55,56 @@ def evaluate_package(question, package)
+  #   towards grade/exp calculation
+  # @return [Array<(Boolean, Integer, Course::Assessment::Answer::ProgrammingAutoGrading), Integer>]
+  #   The correctness apparent to student ('True' if answer passes public and private test
+  #   cases), grade, the programming auto grading record, and the evaluation result's id.
+  def build_result(question, evaluation_result, graded_test_case_types:)
+    auto_grading = build_auto_grading(question, evaluation_result)
+    graded_test_count = question.test_cases.where(test_case_type: graded_test_case_types).size
+    number_correct = auto_grading.test_results.
+                     select { |r| graded_test_case_types.include?(r.test_case.test_case_type) }.
+                     map(&:passed?).count(true)
+
+    considered_correct = check_correctness(question, auto_grading)
+    grade = if graded_test_count == 0
+              question.maximum_grade
+            else
+              question.maximum_grade * number_correct / graded_test_count
+            end
+    [considered_correct, grade, auto_grading, evaluation_result.evaluation_id]
+  end
+
+  # Builds a ProgrammingAutoGrading instance from the question and package evaluation result.
+  #
+  # @param [Course::Assessment::Question::Programming] question The programming question being
+  #   graded.
+  # @param [Course::Assessment::ProgrammingEvaluationService::Result] evaluation_result The
+  #   result of evaluating the package.
+  # @return [Course::Assessment::Answer::ProgrammingAutoGrading] auto_grading The
+  #   ProgrammingAutoGrading instance
+  def build_auto_grading(question, evaluation_result)
     auto_grading = Course::Assessment::Answer::ProgrammingAutoGrading.new(actable: nil)
     set_auto_grading_results(auto_grading, evaluation_result)
     build_test_case_records(question, auto_grading, evaluation_result.test_reports)
-    test_results = auto_grading.test_results
-    test_cases = question.test_cases
-
-    if ignore_evaluation
-      test_results = test_results.reject { |r| r.test_case.evaluation_test? }
-      test_cases = test_cases.reject(&:evaluation_test?)
-    end
-
-    number_correct = test_results.map(&:passed?).count(true)
-    test_count = test_cases.size
+    auto_grading
+  end
 
-    all_correct = number_correct == test_count
-    grade = if test_count == 0
-              question.maximum_grade
-            else
-              question.maximum_grade * number_correct / test_count
-            end
-    [all_correct, grade, auto_grading, evaluation_result.evaluation_id]
+  # Checks if the answer passes all public and private test cases.
+  #
+  # @param [Course::Assessment::Question::Programming] question The programming question being
+  #   graded.
+  # @param [Course::Assessment::Answer::ProgrammingAutoGrading] auto_grading The
+  #   ProgrammingAutoGrading instance
+  # @return [Boolean] True if the evaluated answer passes all public and private test cases
+  def check_correctness(question, auto_grading)
```

Comment:
Add a spec for this. You could create the scenario of passing public and private, but not evaluation, and still being allowed to continue in `spec/features/course/assessment/submission/autograded_spec.rb`

Alternatively, add the test to `spec/services/course/assessment/answer/programming_auto_grading_service_spec.rb`